types/rolling-rate-limiter/index.d.ts

```diff
@@ -0,0 +1,38 @@
+// Type definitions for rolling-rate-limiter 0.1
+// Project: https://github.com/peterkhayes/rolling-rate-limiter
+// Definitions by: Jonas Lochmann <https://github.com/l-jonas>
+// Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
+
+export = RollingRateLimiter;
+
+declare function RollingRateLimiter(options: RollingRateLimiter.InMemoryOptions): RollingRateLimiter.SyncOrAsyncLimiter;
+declare function RollingRateLimiter(options: RollingRateLimiter.WithRedisOptions): RollingRateLimiter.AsyncLimiter;
+
+declare namespace RollingRateLimiter {
+  interface GeneralOptions {
+    interval: number;
+    maxInInterval: number;
+    minDifference?: number;
+  }
+
+  interface RedisOptions {
+    redis: CompatibleRedisClient;
+    namespace?: string;
+  }
+
+  interface CompatibleRedisClient {
+    multi: () => any;
+  }
+
+  type InMemoryOptions = GeneralOptions;
+  type WithRedisOptions = GeneralOptions & RedisOptions;
+
+  type AsyncLimiterWithToken = (token: string, callback: AsyncLimiterCallback) => void;
+  type AsyncLimiterWithoutToken = (callback: AsyncLimiterCallback) => void;
+  type AsyncLimiterCallback = (err: any, timeLeft: number, actionsLeft: number) => void;
+  type AsyncLimiter = AsyncLimiterWithToken & AsyncLimiterWithToken;
+
+  type SyncLimiter = (token?: string) => number;
+
+  type SyncOrAsyncLimiter = SyncLimiter & AsyncLimiter;
```

Comment:
You are naming this as SyncOrAsync but on type side you are doing intersection instead of union..

Reply:
If I replace ``SyncLimiter & AsyncLimiter`` by ``SyncLimiter | AsyncLimiter``, then the test which is based on the Readme of the package does not pass.

Should I rename it to SyncAndAsyncListener? I used "or" because it can be used as sync or async function.

To the background: When the in-memory-storage is used, then the returned function can be used with and without callback (asynchronous or synchronous).